src/librustc/traits/error_reporting.rs

```diff
@@ -551,6 +551,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                         let OnUnimplementedNote { message, label }
                             = self.on_unimplemented_note(trait_ref, obligation);
                         let have_alt_message = message.is_some() || label.is_some();
+                        let span = match self.tcx.sess.codemap().span_to_snippet(span) {
+                            Ok(ref s) if s.starts_with("for ") => {
+                                // On for loops, this error is caused by the element being iterated
```

Comment:
Hmm, I like this span a lot, but I wonder if this is the right way to do this. Maybe we should adjust the span in HIR lowering instead? Roughly here:

https://github.com/rust-lang/rust/blob/a0dcecff90c45ad5d4eb60859e22bb3f1b03842a/src/librustc/hir/lowering.rs#L3167-L3173

Reply:
Reasonable, it's definitely an overuse of `span_until_char`, it was just slightly faster to figure how to do this than the proper way. Should it point at only the iterator expression or at the entire `for _ in iterator_expr`?

Reply:
@estebank good question, I don't know the answer. Just pointing at the span of the iterator expression is obviously *easier*... and maybe better? I guess that showing the `for` may help convey that it's "because of the `for`" that `IntoIterator` is needed... but that seems sort of obvious...

If we do it in lowering, we can also introduce a "synthetic" span, right, that would let us give a very tailored message if we wanted?

Reply:
> If we do it in lowering, we can also introduce a "synthetic" span, right, that would let us give a very tailored message if we wanted?

I think so. If we want to point at the `for`, we could just do that, which would look better ni the case where `for` and `expr` are not in the same line:

```
X |    for _ in expr {
  |    ---      ^^^^ `&str` is not an iterator
  |    |
  |    needed because of this `for` loop
```

Reply:
oooh, that's nice

Reply:
I looked at this and I have to rework the code to not lose the for-loop's context by the time the trait is being resolved. It might take me a while.

Reply:
Modified the expansion code to use the appropriate span, but that makes it harder to add the appropriate span under `for`.